components/global/product_switcher_menu/product_switcher_menu.tsx

```diff
@@ -81,35 +83,39 @@ class ProductSwitcherMenu extends React.PureComponent<Props> {
                             }
                         />
                     </SystemPermissionGate>
-                    <Menu.ItemLink
-                        id='integrations'
-                        show={showIntegrations}
-                        to={'/' + this.props.teamName + '/integrations'}
-                        text={formatMessage({id: 'navbar_dropdown.integrations', defaultMessage: 'Integrations'})}
-                        icon={
-                            <Icon
-                                size={16}
-                                glyph={'webhook-incoming'}
-                            />
-                        }
-                    />
-                    <TeamPermissionGate
-                        teamId={this.props.teamId}
-                        permissions={[Permissions.SYSCONSOLE_WRITE_PLUGINS]}
-                    >
-                        <Menu.ItemToggleModalRedux
-                            id='marketplaceModal'
-                            modalId={ModalIdentifiers.PLUGIN_MARKETPLACE}
-                            show={!this.props.isMobile && this.props.enablePluginMarketplace}
-                            dialogType={MarketplaceModal}
-                            text={formatMessage({id: 'navbar_dropdown.marketplace', defaultMessage: 'Marketplace'})}
+                    {isMessaging && (
+                        <Menu.ItemLink
+                            id='integrations'
+                            show={showIntegrations}
```

Comment:
Looks like the existing menu item component has a `show` prop that is currently used to indicate whether the menu item should be rendered or not. Thoughts on using the prop instead for both menu items?

```suggestion
                        <Menu.ItemLink
                            id='integrations'
                            show={isMessaging && showIntegrations}
```

Reply:
I am typically against the usage of a prop for this, since it will call the function regardless. With the approach I took it won't do that. It is just a small performance improvement.

Reply:
@deanwhillier I just took a look at the implementation of Menu.ItemLink and IMO it massively over-engineered (several nested functions and classes). All the more reason not even calling.

Reply:
Actually, MenuItemLink is build with a HOC at load time, so it is not generating that many calls, actually if show is false, it is not rendering anything, or calling any other internal function, so should be pretty efficient.

Said that, you are right, the menu system has some over-engineering, the plan was to reduce the code repetition of each item using HOC that defines what a menu item is, and nesting the items inside the `Menu` component to give a "namespace" to use them.

Reply:
@jespino as far as I can see it, the `MenuItemLink` is being called from `Menu` (namespace). The actual `MenuItemLink` consists of a function that takes in the react-router `Link` component and passes it into a function, that creates a new class. Inside that classes render function is the condition to return `null` if `show` is false. All of this can be prevented by just not calling it! :D

Reply:
You can prevent some things not calling it, but you are assuming that `MenuItemLink` is calling the `menuItem` HOC funtion, that creates the class, but the reality is `MenuItemLink` is a class that is generated on load time (whenever you load the javascript) calling the `menuItem` HOC function, on top of the `MenuItemLinkImpl`. That means, as soon as you loaded the javascript `MenuItemLink` is an instance of a class that looks like this:

```ts
      class MenuItem extends React.PureComponent<Props & React.ComponentProps<typeof Component>> {
          public static defaultProps = {
              show: true,
          };
   
          public static displayName?: string;
   
          public render() {
              const {id, show, icon, text, to} = this.props;
              if (!show) {
                  return null;
              }
   
              let textProp: React.ReactNode = text;
              if (icon) {
                  textProp = (
                      <React.Fragment>
                          <span className='icon'>{icon}</span>
                          {text}                           
                      </React.Fragment>
                  );
              }                              
   
              return (
                  <li
                      className={classNames('MenuItem', {
                          'MenuItem--with-icon': icon,
                      })}
                      role='menuitem'
                      id={id}
                  >
                      <Link to={to}><span className='MenuItem__primary-text'>{text}</span></Link>
                  </li>
              );
        }
    }
```

So all that calls that you are talking about, are just reduced to the `render` function (and the lifecycle functions of the component).

You can read more about HOC here: https://reactjs.org/docs/higher-order-components.html

Reply:
I know what a HOC is and I know how to improve performance of React applications (done that several times) and regardless of when the class instance gets called React will go and try to mount the component (even if it eventually is `null`).

That is what I am trying to prevent. I know perfectly well, that this might be just a tiny improvement, but I have seen several apps in the past that were pretty inefficient just because of complexity and small things like these. "Death by a thousand cuts".

If you don't like the approach: fine. I can change that back.